src/core/common/message.cpp

```diff
@@ -447,138 +449,118 @@ uint16_t Message::Read(uint16_t aOffset, uint16_t aLength, void *aBuf) const
-        bytesCopied += bytesToCopy;
-        aBuf = static_cast<uint8_t *>(aBuf) + bytesToCopy;
+    // Special case for the first buffer
 
-        aOffset = 0;
-    }
-    else
+    if (aOffset < kHeadBufferDataSize)
     {
-        aOffset -= kHeadBufferDataSize;
+        aChunkData   = GetFirstData() + aOffset;
+        aChunkLength = kHeadBufferDataSize - aOffset;
+        ExitNow();
     }
 
-    // advance to offset
-    curBuffer = GetNextBuffer();
+    aOffset -= kHeadBufferDataSize;
 
-    while (aOffset >= kBufferDataSize)
-    {
-        OT_ASSERT(curBuffer != nullptr);
+    // Find the `Buffer` matching the offset
 
-        curBuffer = curBuffer->GetNextBuffer();
-        aOffset -= kBufferDataSize;
-    }
-
-    // begin copy
-    while (aLength > 0)
+    while (true)
     {
-        OT_ASSERT(curBuffer != nullptr);
+        aChunkBuffer = aChunkBuffer->GetNextBuffer();
+        OT_ASSERT(aChunkBuffer != nullptr);
 
-        bytesToCopy = kBufferDataSize - aOffset;
-
-        if (bytesToCopy > aLength)
+        if (aOffset < kBufferDataSize)
         {
-            bytesToCopy = aLength;
+            aChunkData   = aChunkBuffer->GetData() + aOffset;
+            aChunkLength = kBufferDataSize - aOffset;
+            ExitNow();
         }
 
-        memcpy(aBuf, curBuffer->GetData() + aOffset, bytesToCopy);
-
-        aLength -= bytesToCopy;
-        bytesCopied += bytesToCopy;
-        aBuf = static_cast<uint8_t *>(aBuf) + bytesToCopy;
-
-        curBuffer = curBuffer->GetNextBuffer();
-        aOffset   = 0;
+        aOffset -= kBufferDataSize;
     }
 
 exit:
-    return bytesCopied;
+    if (aChunkLength > aLength)
+    {
+        aChunkLength = aLength;
+    }
+
+    aLength -= aChunkLength;
 }
 
-int Message::Write(uint16_t aOffset, uint16_t aLength, const void *aBuf)
+void Message::GetNextChunk(uint16_t &      aLength,
+                           const Buffer *& aChunkBuffer,
+                           const uint8_t *&aChunkData,
+                           uint16_t &      aChunkLength) const
 {
-    Buffer * curBuffer;
-    uint16_t bytesCopied = 0;
-    uint16_t bytesToCopy;
+    // This method gets the next message data chunk. On input, the
+    // `aChunkBuffer` should point to the `Buffer` containing the
+    // previous chunk. On exit, it is updated to point to  the
+    // `Buffer` containing the next chunk, `aChunkData` and
+    // `aChunkLength` are also updated and `aLength` is decreased.
+    // `aChunkLength` is set to zero when there is no more byte to
+    // read (all bytes are read and `aLength` is zero on input).
 
-    OT_ASSERT(aOffset + aLength <= GetLength());
+    VerifyOrExit(aLength > 0, aChunkLength = 0);
 
-    if (aOffset + aLength >= GetLength())
-    {
-        aLength = GetLength() - aOffset;
-    }
+    aChunkBuffer = aChunkBuffer->GetNextBuffer();
+    OT_ASSERT(aChunkBuffer != nullptr);
 
-    aOffset += GetReserved();
+    aChunkData   = aChunkBuffer->GetData();
```

Comment:
I think in `GetNextChunk`, the outcome `aChunkData` actually always points to the start of `aChunkBuffer`. This function  doesn't actually depend on `aChunkData` of the previous chunk. Maybe we can remove this parameter.

Reply:
The primary goal of adding the `GetFirst/NextChunk()` is to help encapsulate all the common code that is used by `Read()`, `Write()` and `UpdateChecksum()` into the same helper method (so we don't repeat same code). 

They all need to get the pointer to the start of the chunk data (from the related `Buffer`), so I think it makes sense for this to be part of the common helper method itself (than each of them doing the same).

Also keeping `aChunkData` as a parameter helps keep the behavior of `GetFirstChunk()` and `GetNextChunk()` to be similar.

Reply:
Make sense.